metascore/main.mo

```diff
@@ -129,51 +150,123 @@ shared ({caller = owner}) actor class Metascore() : async M.FullInterface {
+                    };
+                };
+            };
+            case (? gr) {
+                let ps = gr.players;
+                for (p in players.vals()) {
+                    let pID = p.player;
+                    gr.players.put(pID, p);
+                };
+                state.games.put(gameID, {
+                    metadata = gr.metadata;
+                    players  = ps;
+                });
+                for ((_, p) in ps.entries()) {
+                    let pID = p.player;
+                    switch (scores.get(pID)) {
+                        case (null) {
+                            let ps = GameRecord.emptyPlayerGameScores(1);
+                            ps.put(gameID, state.getMetascore(gameID, pID));
+                            scores.put(pID, ps);
+                        };
+                        case (? ps) {
+                            ps.put(gameID, state.getMetascore(gameID, pID));
+                            scores.put(pID, ps);
+                        };
+                    };
+                };
+             };
+        };
+    };
+
     private func queryAllScores() : async () {
-        Debug.print("Getting scores...");
         for ((gID, g) in state.games.entries()) {
             let scores = await getScores(gID);
-            state.putGameRecord(gID, g.metadata, scores);
+            putGameRecord(gID, g.metadata, scores);
         };
     };
 
-    private func getScores(id : MPublic.GamePrincipal) : async [MPublic.Score] {
+    private func getScores(id : MPublic.GamePrincipal) : async [GameRecord.PlayerRecord] {
         let game : MPublic.GameInterface = actor(Principal.toText(id));
-        Array.sort(
-            await game.metascoreScores(),
+        sortScores(await game.metascoreScores());
+    };
+
+    private func sortScores(scores : [MPublic.Score]) : [GameRecord.PlayerRecord] {
+        Array.sort<GameRecord.PlayerRecord>(
+            Array.map<MPublic.Score, GameRecord.PlayerRecord>(
+                scores,
+                func ((player, score) : MPublic.Score) : GameRecord.PlayerRecord {
+                    { player; score; };
+                },
+            ),
             // Sort from high to low.
-            func (a : MPublic.Score, b : MPublic.Score) : Order.Order {
-                let (x, y) = (a.1, b.1);
+            func (a : GameRecord.PlayerRecord, b : GameRecord.PlayerRecord) : Order.Order {
+                let (x, y) = (a.score, b.score);
                 if      (x < y)  { #greater; }
                 else if (x == y) { #equal;   }
                 else             { #less;    };
             },
         );
     };
 
+    public query func getTop(n : Nat) : async [MPublic.Score] {
+        var top : [MPublic.Score] = [];
+        for ((p, scores) in scores.entries()) {
+            top := Array.append<MPublic.Score>(top, [(
+                p, GameRecord.totalScore(scores)
+            )]);
+        };
+        top;
+    };
+
```

Comment:
This is a great start. Obviously we will need to working more on our query methods when we start building the frontend.